File: 한승우/10월 3주차/도둑질.cpp
#include <string>
#include <vector>

using namespace std;

/*
1. 완탐
100만 C 50만 * 100만 = 시간 초과

2. DP
홀수인 상태에서 문제네.

선형적으로 메모이제이션할 수 없는 것 같은데?
짝수일 때는 선형적으로 가능
홀수일 때는 불가능

경우를 더 나누면 가능

1. i번째 집을 선택했을 때의 최댓값
2. 첫번째 집을 선택하는 경우 (idx : 1 ~ n - 1)
dp[i] = max(dp[i - 1], dp[i - 2] + house[i]);

마지막 집을 선택하는 경우 (idx : 2 ~ n)

3. 0, 1 : 첫번째 집, 2 : 1과 2중에 큰 집
*/

int solution(vector<int> money) {
    int answer = 0;
    vector<int> dp(money.size(), 0);
    
    //1. 첫번째 집 선택
    dp[0] = money[0];
    dp[1] = max(money[0], money[1]);
    for(int i = 2; i < dp.size() - 1; i++){
        dp[i] = max(dp[i - 1], dp[i - 2] + money[i]);
    }
    
    answer = max(answer, dp[money.size() - 2]);
    
    //2. 마지막 집 선택
    dp[1] = money[1];
    dp[2] = max(money[1], money[2]);
    for(int i = 3; i < dp.size(); i++){
        dp[i] = max(dp[i - 1], dp[i - 2] + money[i]);
    }
    
    answer = max(answer, dp[money.size() - 1]);
    
    return answer;
}

File: 송하연/10월 3주차/도둑질.java
class Solution {
    public int solution(int[] money) {
        int answer = 0;
        
        int n=money.length;
        
        //첫 번째 집 털었을 때 마지막 집 털지 X
        int[] dp1 = new int[n];
        dp1[0] = money[0]; //max(1)
        dp1[1] = Math.max(money[0],money[1]); //max(1,2)
        for(int i=2; i<n-1; i++){
            dp1[i] = Math.max(dp1[i-1], dp1[i-2]+money[i]); //max(2,1+3)
        }
        
        //첫 번째 집 털지 X 마지막 집 털기
        int[] dp2 = new int[n];
        dp2[0] = 0; 
        dp2[1] = money[1];
        for(int j=2; j<n; j++){
            dp2[j] = Math.max(dp2[j-1], dp2[j-2]+money[j]);
        }
        
        return Math.max(dp1[n-2],dp2[n-1]);

    }
}
